2016/presentations/Python_for_Networking_Devices/text.md

```diff
@@ -0,0 +1,139 @@
+Arista device:
+```
+Address         Age (min)  Hardware Addr   Interface
+10.220.88.1             0  001f.9e92.16fb  Vlan1, Ethernet1
+10.220.88.21            0  1c6a.7aaf.576c  Vlan1, not learned
+10.220.88.28            0  5254.00ee.446c  Vlan1, not learned
+10.220.88.29            0  5254.0098.69b6  Vlan1, not learned
+10.220.88.30            0  5254.0092.13bb  Vlan1, not learned
+10.220.88.38            0  0001.00ff.0001  Vlan1, not learned
+```
+
+On a Juniper device, it will look more like this:
+```
+MAC Address       Address         Name                      Interface           Flags
+00:1f:9e:92:16:fb 10.220.88.1     10.220.88.1               vlan.0              none
+00:19:e8:45:ce:80 10.220.88.22    10.220.88.22              vlan.0              none
+f0:ad:4e:01:d9:33 10.220.88.100   10.220.88.100             vlan.0              none
+Total entries: 3
+```
+
+Not only can the show commands to execute on the device differ per vendor, they also provide different text format
+output, which in case of CLI access has to be parsed individually. Netconf like access methods support the retrieval
+of structured data, which is slightly better, but this still doesn’t cover fields not provided in one vendor vs.
+another (like for example the lack of an age timer in the Juniper output above).
+
+##Generic Access Libraries
+
+To access devices directly via their SSH or Netconf interface, generic python libraries such as
+[pexpect](https://github.com/pexpect/pexpect), [paramiko](https://github.com/paramiko/paramiko) and
+[ncclient](https://github.com/ncclient/ncclient) can be used. They allow for programmatic access from scripts
+(for network engineers who are trying to figure out how to code ;) ) but don't provide any ease of use in terms of
+vendor specificity. You will still have to deal with the little differences the vendors bring, different login
+procedured or additional escape chars). You will take care of different lines of configuration or show commands you
+need to pass into paramiko or pexpect for each device or vendor, or with the varying support of netconf on each
+device for ncclient.
+
+##Specific Network Vendor Libraries
+
+Since Python became more and more popular amongst network engineers, network vendors started to provide their own
+libraries to facilitate device interaction - to mention a few: 
+
+* Arista's [pyeapi](https://github.com/arista-eosplus/pyeapi)
+* Cisco IOS-XR [pyiosxr](https://github.com/fooelisa/pyiosxr)
+* Juniper's [py-junos-eznc](https://github.com/Juniper/py-junos-eznc )
+
+Each of them supports their specific device with its specific capabilities. Juniper's py-junos-eznc  provides access
+via Netconf, whereas pyiosxr 'mimics' netconf support (which is not directly available) via SSH (pexpect). They ease
+config operations on the devices, such as replacing or merging configurations in one go, or . Each of them very
```

Comment:
This sentence doesn't make sense. Hard to guess what the author wanted to say. Needs rephrasing.